Add AWS Serverless Forwarder to Obs Guide (#2214)
elastic/observability-docs: docs/en/observability/aws-elastic-serverless-forwarder.asciidoc

:aws: AWS

[[aws-elastic-serverless-forwarder]]
= Elastic Serverless Forwarder

// should we expand titleabbrev to AWS or Amazon Web Services Elastic Serverless Forwarder? What is correct nomenclature?

++++
<titleabbrev>AWS Elastic Serverless Forwarder</titleabbrev>

> ++++

Reviewer: Colleen's suggestion is inline with their trademark guidelines.

> ++++

Reviewer: Nothing here that titleabbrev is also used for generated link text.

> ++++

Author: > Or if the name of the product is "Elastic Serverless Forwarder" we could use something like "Elastic Serverless Forwarder for Amazon Web Services (AWS)".

We definitely need clarity on naming and how we connect it to AWS. I like this suggestion as it handles possible developments like Elastic Serverless Forwarder for BlahBlah (I know that other cloud platforms implement differently but it's not hard to imagine same concept being used for other cloud providers).

As we don't have the **Cloud observability** section yet, and the AWS bit is needed to make the context clear, I'd experimented with using titleabbrev to see how the menu item looks. I could see us having a longer page title and filenames but using the abbrev to shorten (e.g. AWS rather than Amazon Web Services). What do you think to the naming point @ravikesarwani?

> ++++

Author: Any thoughts on the naming issue @ravikesarwani ?

> ++++

Reviewer 2: Using AWS in the titleabbrev is fine, so something like "Elastic Serverless Forwarder for AWS" should be okay from my perspective (trying to keep it smaller)

:keywords: serverless
:description: The Elastic Serverless Forwarder is an Amazon Web Services ({aws}) Lambda function that ships logs from your {aws} environment to Elastic.

The Elastic Serverless Forwarder is an Amazon Web Services ({aws}) Lambda function that ships logs from your {aws} environment to Elastic.

[discrete]
[[aws-serverless-forwarder-overview]]
= Overview
The Elastic Serverless Forwarder can forward {aws} data to self-managed or cloud-hosted Elastic environments. It supports the following inputs:

> - Amazon S3 SQS (Simple Queue Service) event notifications

Reviewer 2: Amazon S3 (via SQS event notifications)

> - Amazon S3 (via SQS (Simple Queue Service) event notifications)

Author: Think I was trying to do too much with this bullet - going to simplify the bullet text and explain both terms in the line 47 section. The terms are familiar enough with Amazon users that we can give this first reference a pass.

- Amazon Kinesis Data Streams
- Amazon CloudWatch Logs subscription filters

> - Direct Amazon SQS message payload

Reviewer 2: Amazon SQS message payload

[role="screenshot"]
image::aws-serverless-lambda-flow.png[AWS Lambda flow]

> The Lambda deployment sets up an SQS queue automatically. This is used to trigger a new function invocation that Lambda uses to continue from exactly where the last function run was terminated. By default a Lambda function runs for a maximum of 15 minutes. When processing event data there’s a possibility that the function may be exited by {aws} in the middle of processing. The function handles this scenario gracefully by keeping track of the last offset processed.

Reviewer 2: The Lambda deployment sets up a continuing SQS queue automatically.

Author: Think I edited that original reference out lol - what do you mean by "continuing"?
I don't think Amazon refer to a "continuing SQS queue" in their docs so trying to understand what this means in this context e.g. "persistent"?

Reviewer 2: Yes, its an SQS queue but we call it continuing queue (our name). This is what we are showing in our diagram as well.
This helps us to differentiate and refer to the specific queue (as you can see there are many) by its purpose (helps with checkpointing and continue where where we left of in the last run).

Author: Ah, thanks - I see it in the diagram now. Updating locally, will push another commit and then ask for a peer review from an Obs writer.

> The Lambda deployment sets up an SQS _Continuing queue_ automatically. This is used to trigger a new function invocation that Lambda uses to continue from exactly where the last function run was terminated. By default a Lambda function runs for a maximum of 15 minutes. When processing event data there’s a possibility that the function may be exited by {aws} in the middle of processing. The function handles this scenario gracefully by keeping track of the last offset processed.

Author: Thanks, I like your new suggestion

Any exception or fail scenarios related to ingestion are handled by Lambda gracefully using a _Replay queue_. Data in the replay queue is stored as individual events. Lambda keeps track of any failed events and writes them to a replay queue that can then be consumed by adding an additional SQS trigger via Lambda.

> You can use the <<sample-s3-config-file,config.yaml>> file to configure the service for each input type, including information such as SQS queue ARN and {es} connection details. You can create multiple input sections within the configuration file to map different inputs to specific log types.

Reviewer 3: >SQS queue ARN

I don't know what this means. Do we want to assume the reader will know or should we explain?

Reviewer: AWS users will know what this means.

> The Lambda function also supports writing directly to an index, alias, or custom data stream. This enables existing {es} users to re-use index templates, ingest pipelines, or dashboards that are already created and connected to business processes.

Reviewer 3: I don't really know what "are connected to business processes" means. Does that just mean they are already in use? 

```suggestion
The Lambda function also supports writing directly to an index, alias, or custom data stream. This enables existing {es} users to re-use index templates, ingest pipelines, or dashboards that already exist and are connected to business processes.
```

Author: Yes but I agree some examples or details would help here... @ravikesarwani could you help with this one?

> The best way to get started is to install the appropriate integration assets via {kib}. These integration assets include pre-built dashboards, ingest node configurations, and other assets that help you get the most out of the data you ingest. The integrations use https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html[data streams] with specific https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme[naming conventions] that provide you with more granular controls and flexibility on managing data ingestion.

Reviewer 3: Should this be with the `= Get started` section below?

Author: Yeah, it is a little out of place / overlapping with either the Get started _section_ or the Deploy topic itself - I left it in as it was a useful CTA inside quite a dense introductory section but I'll try moving it into Deploy I think...

The best way to get started is to install the appropriate https://docs.elastic.co/en/integrations[integrations] via {kib}. These integrations include pre-built dashboards, ingest node configurations, and other assets that help you get the most out of the data you ingest. The integrations use https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html[data streams] with specific https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme[naming conventions] that provide you with more granular controls and flexibility on managing data ingestion.

> 

Reviewer 2: Add assets after integration.
The best way to get started is to install the appropriate https://docs.elastic.co/en/integrations[integrations] **assets** via {kib}. These integrations **assets** include pre-built dashboards, ingest node configurations, and other assets that help you get the most out of the data you ingest. The integrations use https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html[data streams] with specific https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme[naming conventions] that provide you with more granular controls and flexibility on managing data ingestion.

[discrete]
[[aws-serverless-forwarder-inputs]]
= Inputs

[discrete]
[[aws-serverless-forwarder-inputs-direct]]
== Direct Amazon SQS message payload

> 

Reviewer 2: We can get rid of "Direct" and keep it simple as "Amazon SQS message payload".
Can we also change the order to match the top and the same exact names. Order: Amazon S3=> Kinesis => cloudwatch log =>SQS.

For other log types, you can optionally set the `es_datastream_name` value in the configuration file according to the naming convention of the {es} data stream and integration.  If the `es_datastream_name` is not specified, and the log cannot be matched with any of the above {aws} services, then the dataset will be set to `generic` and the namespace set to `default`, pointing to the data stream name `logs-generic-default`.

For more information on creating SQS event notifications for S3 buckets, read the https://docs.aws.amazon.com/AmazonS3/latest/userguide/ways-to-add-notification-config-to-bucket.html[{aws} documentation].

NOTE: You must set a visibility timeout of `910` seconds for any SQS queues you want to use as a trigger. This is 10 seconds greater than the Elastic Serverless Forwarder Lambda timeout.

[discrete]
[[aws-serverless-forwarder-inputs-kinesis]]
== Amazon Kinesis Data Streams

> The Lambda function can ingest logs contained in the payload of a Kinesis Data Stream record and send them to Elastic. The Kinesis Data Stream serves as a trigger for the Lambda function. When a new record gets written to a Kinesis Data Stream, the Lambda function triggers.

Reviewer 3: ```suggestion
The Lambda function can ingest logs contained in the payload of a Kinesis Data Stream record and send them to Elastic. The Kinesis Data Stream serves as a trigger for the Lambda function. When a new record gets written to a Kinesis Data Stream, it triggers the Lambda function.
```

You can set up separate Kinesis Data Streams for each type of log. The `es_datastream_name` parameter in the config file is mandatory. If this value is set to an {es} data stream, the type of log must be correctly defined with configuration parameters. A single configuration file can have many input sections, pointing to different data streams that match specific log types.

[discrete]
[[aws-serverless-forwarder-inputs-cloudwatch]]
== Amazon CloudWatch Logs subscription filters

> The Lambda function can ingest logs contained in the message payload of CloudWatch Logs events and send them to Elastic. The CloudWatch Logs service serves as a trigger for the Lambda function. When a new record gets written to a Kinesis data stream, the Lambda function triggers.

Reviewer 3: ```suggestion
The Lambda function can ingest logs contained in the message payload of CloudWatch Logs events and send them to Elastic. The CloudWatch Logs service serves as a trigger for the Lambda function. When a new record gets written to a Kinesis data stream, it triggers the Lambda function.
```

> The Lambda function can ingest logs contained in the message payload of CloudWatch Logs events and send them to Elastic. The CloudWatch Logs service serves as a trigger for the Lambda function. When a new record gets written to a Kinesis data stream, the Lambda function triggers.

Reviewer 4: beware we are mixing cloudwatch logs and kinesis data stream here in the description:
```suggestion
The Lambda function can ingest logs contained in the message payload of CloudWatch Logs events and send them to Elastic. The CloudWatch Logs service serves as a trigger for the Lambda function. When a new event gets written to a CloudWatch Logs log stream, it triggers the Lambda function.
```

about:
> it triggers the Lambda function

indeed it's a little different than this. multiple events written to the log stream can trigger the lambda function "in batch".

You can set up separate CloudWatch Logs groups for each type of log. The `es_datastream_name` parameter in the config file is mandatory. If this value is set to an {es} data stream, the type of log must be correctly defined with configuration parameters. A single configuration file can have many input sections, pointing to different CloudWatch Logs groups that match specific log types.

[discrete]
[[aws-serverless-forwarder-inputs-direct]]
== Amazon SQS message payload

> The Lambda function can ingest logs contained within the payload of an Amazon SQS body record and send them to Elastic. The SQS queue serves as a trigger for the Lambda function. When a new record gets written to an SQS queue, the Lambda function triggers.

Reviewer 3: ```suggestion
The Lambda function can ingest logs contained within the payload of an Amazon SQS body record and send them to Elastic. The SQS queue serves as a trigger for the Lambda function. When a new record gets written to an SQS queue, it triggers the Lambda function.
```
